dbms/src/Operators/AutoPassThroughHashAggContext.h

```diff
@@ -0,0 +1,225 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Interpreters/Aggregator.h>
+#include <Operators/AutoPassThroughHashAggHelper.h>
+#include <tipb/executor.pb.h>
+
+namespace DB
+{
+struct AutoPassThroughSwitcher
+{
+    explicit AutoPassThroughSwitcher(const ::tipb::Aggregation & aggregation)
+    {
+        if (aggregation.has_pre_agg_mode())
+        {
+            has_set = true;
+            mode = aggregation.pre_agg_mode();
+        }
+    }
+    AutoPassThroughSwitcher(bool has_set_, ::tipb::TiFlashPreAggMode mode_)
+        : has_set(has_set_)
+        , mode(mode_)
+    {}
+
+    bool enable() const { return has_set && !forcePreAgg(); }
+
+    bool forcePreAgg() const { return mode == ::tipb::TiFlashPreAggMode::ForcePreAgg; }
+
+    bool forceStreaming() const { return mode == ::tipb::TiFlashPreAggMode::ForceStreaming; }
+
+    bool isAuto() const { return mode == ::tipb::TiFlashPreAggMode::Auto; }
+
+    bool has_set = false;
+    ::tipb::TiFlashPreAggMode mode;
+};
+
+class AutoPassThroughHashAggContext
+{
+public:
+    AutoPassThroughHashAggContext(
+        const Block & child_header_,
+        const Aggregator::Params & params_,
+        Aggregator::CancellationHook && hook,
+        const String & req_id_,
+        UInt64 row_limit_unit_,
+        UInt64 adjust_state_unit_num = 1,
+        UInt64 other_state_unit_num = 5)
+        : state(State::Init)
+        , many_data(std::vector<AggregatedDataVariantsPtr>(1, nullptr))
+        , adjust_row_limit(row_limit_unit_ * adjust_state_unit_num)
+        , other_state_row_limit(row_limit_unit_ * other_state_unit_num)
+        , row_limit_unit(row_limit_unit_)
+        , log(Logger::get(req_id_))
+    {
+        aggregator = std::make_unique<Aggregator>(params_, req_id_, 1, nullptr);
+        aggregator->setCancellationHook(hook);
+        aggregator->initThresholdByAggregatedDataVariantsSize(1);
+        many_data[0] = std::make_shared<AggregatedDataVariants>();
+        agg_process_info = std::make_unique<Aggregator::AggProcessInfo>(aggregator.get());
+        RUNTIME_CHECK(adjust_row_limit > 1024 && other_state_row_limit > 1024);
+        RUNTIME_CHECK(aggregator->getParams().keys_size > 0);
+
+        auto header = aggregator->getHeader(/*final=*/true);
+        const auto & aggregate_descriptions = aggregator->getParams().aggregates;
+        column_generators = setupAutoPassThroughColumnGenerator(header, child_header_, aggregate_descriptions, log);
+        RUNTIME_CHECK(header.columns() == column_generators.size());
+    }
+
+    ~AutoPassThroughHashAggContext() { statistics.log(log); }
+
+    template <bool force_streaming>
+    void onBlock(Block & block)
+    {
+        if constexpr (force_streaming)
+        {
+            statistics.update(State::PassThrough, block.rows());
+            onBlockForceStreaming(block);
+        }
+        else
+        {
+            forceState();
```

Comment:
What does this function do? Could you describe how can this new feature be adaptive for spilling?

Reply:
Once hash table is big enough and it needs spill, will pass through all remaining blocks. No block will be inserted into hash table anymore